### PR TITLE
Add LoaderAllocator as a dac/sos concept

### PR DESCRIPTION
**Pull request by leculver**
- DacpModuleData: Repurposed two unused fields to produce the module's LoaderAllocator and ThunkHeap.  The LoaderAllocator often will be the same as the SystemDomain's, except in the case of collectable assemblies.
- Added the way to get the direct LoaderAllocator address from an AppDomain.
- Added a new way to enumerate the heaps associated with a LoaderAllocator in a future-proof way.  Now when the implementation of CLR changes to add, remove, or rename/repurpose an existing heap, we can still produce the right diagnostic data without needing to ship a new interface.

This does expose LoaderAllocators as a new concept to the dac, but the original design of hiding them actually causes a lot more problems than it solves.

For example, the SystemDomain and AppDomain share the same LoaderAllocator and associated heaps...but that implementation detail is opaque to something like ClrMD or SOS (and was different on desktop CLR, so it's tough to take a dependency on that detail).  This means things like !eeheap would dutifully double report heap ranges and sizes.  We are now able to distinguish when we've already seen the same LoaderAllocator and take appropriate action.

We also now need to know AppDomain's LoaderAllocator address to compare to DacpModuleData's to see if we should enumerate that module's heaps or not.

Lastly, the interface was written so that it could be backported to Desktop CLR if we ever deem that necessary.  That's why we have a function to get any AppDomain's LoaderAllocator, even though in .Net Core we only have one.  It's unlikely we'll ever backport it, but we do have that option.

### CODE REVIEW

**Review by mikem8361 — APPROVED** [review]

File: src/coreclr/debug/daccess/request.cpp [hunk 1]
```diff
@@ -3551,6 +3553,118 @@ ClrDataAccess::TraverseVirtCallStubHeap(CLRDATA_ADDRESS pAppDomain, VCSHeapType
...
+    "NewStubPrecodeHeap",
+    "IndcellHeap",
+    "LookupHeap",
+    "ResolveHeap",
+    "DispatchHeap",
+    "CacheEntryHeap",
+    "VtableHeap",
+};
+
+const int LoaderHeapCount = 12;
```
Comment (mikem8361) [inline src/coreclr/debug/daccess/request.cpp:3593]: I don't understand the details of the loader heap, but is hard coding the number of heaps the only way? What if it changes?  Is there any static_assert or something to at least make devs aware of this?

Comment (leculver) [reply src/coreclr/debug/daccess/request.cpp:3593]: The value itself isn't fixed.  We can add more entries to `LoaderAllocatorLoaderHeapNames` by increasing `LoaderHeapCount` and adding more (or less) heaps.  It's actually only determining the length of the above array.  I really want to use `_countof(LoaderAllocatorLoaderHeapNames)` below, but `_countof` is only available in MSVC compilers, so instead we have to make a constant to use throughout the code that needs to know the length of that array.


I could make this more obvious by using the value like this:

``` C++
const int LoaderHeapCount = 12;
static const char *LoaderAllocatorLoaderHeapNames[LoaderHeapCount] = 
```

I'll push that change.

Comment (leculver) [reply src/coreclr/debug/daccess/request.cpp]: See: https://github.com/dotnet/runtime/pull/82437/commits/301089ce7dec09e9233592ed3f19f5bf322e1fc8.

Comment (mikem8361) [reply src/coreclr/debug/daccess/request.cpp]: Looks good.

**Comment by mikem8361** [comment]
Looks good.  Thanks.

File: src/coreclr/debug/daccess/request.cpp (hunk 1 above)
Comment (hoyosjs) [reply src/coreclr/debug/daccess/request.cpp]: Could this be sizeof(LoaderAllocatorLoaderHeapNames)/sizeof(const char *)?

Comment (leculver) [reply src/coreclr/debug/daccess/request.cpp]: It's now `ARRAY_SIZE`, I couldn't seem to find that macro when I was using `_countof`.

**Pull request merged**